File: distribution/python/dqops/client/models/check_definition_model.py
from typing import Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

T = TypeVar("T", bound="CheckDefinitionModel")


@_attrs_define
class CheckDefinitionModel:
    """Data quality check definition model

    Attributes:
        check_name (Union[Unset, str]): Check name
        sensor_name (Union[Unset, str]): Sensor name
        rule_name (Union[Unset, str]): Rule name
        help_text (Union[Unset, str]): Help text that is shown in the check editor that describes the purpose and usage
            of the check
        friendly_name (Union[Unset, str]): An alternative check's name that is shown on the check editor.
        standard (Union[Unset, bool]): This is a standard data quality check that is always shown on the data quality
            checks editor screen. Non-standard data quality checks (when the value is false) are advanced checks that are
            shown when the user decides to expand the list of checks.
        custom (Union[Unset, bool]): This check has is a custom check or was customized by the user.
        built_in (Union[Unset, bool]): This check is provided with DQOps as a built-in check.
        can_edit (Union[Unset, bool]): Boolean flag that decides if the current user can update or delete this object.
        yaml_parsing_error (Union[Unset, str]): Optional parsing error that was captured when parsing the YAML file.
            This field is null when the YAML file is valid. If an error was captured, this field returns the file parsing
            error message and the file location.
    """

    check_name: Union[Unset, str] = UNSET
    sensor_name: Union[Unset, str] = UNSET
    rule_name: Union[Unset, str] = UNSET
    help_text: Union[Unset, str] = UNSET
    friendly_name: Union[Unset, str] = UNSET
    standard: Union[Unset, bool] = UNSET
    custom: Union[Unset, bool] = UNSET
    built_in: Union[Unset, bool] = UNSET
    can_edit: Union[Unset, bool] = UNSET
    yaml_parsing_error: Union[Unset, str] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        check_name = self.check_name
        sensor_name = self.sensor_name
        rule_name = self.rule_name
        help_text = self.help_text
        friendly_name = self.friendly_name
        standard = self.standard
        custom = self.custom
        built_in = self.built_in
        can_edit = self.can_edit
        yaml_parsing_error = self.yaml_parsing_error

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if check_name is not UNSET:
            field_dict["check_name"] = check_name
        if sensor_name is not UNSET:
            field_dict["sensor_name"] = sensor_name
        if rule_name is not UNSET:
            field_dict["rule_name"] = rule_name
        if help_text is not UNSET:
            field_dict["help_text"] = help_text
        if friendly_name is not UNSET:
            field_dict["friendly_name"] = friendly_name
        if standard is not UNSET:
            field_dict["standard"] = standard
        if custom is not UNSET:
            field_dict["custom"] = custom
        if built_in is not UNSET:
            field_dict["built_in"] = built_in
        if can_edit is not UNSET:
            field_dict["can_edit"] = can_edit
        if yaml_parsing_error is not UNSET:
            field_dict["yaml_parsing_error"] = yaml_parsing_error

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        d = src_dict.copy()
        check_name = d.pop("check_name", UNSET)

        sensor_name = d.pop("sensor_name", UNSET)

        rule_name = d.pop("rule_name", UNSET)

        help_text = d.pop("help_text", UNSET)

        friendly_name = d.pop("friendly_name", UNSET)

        standard = d.pop("standard", UNSET)

        custom = d.pop("custom", UNSET)

        built_in = d.pop("built_in", UNSET)

        can_edit = d.pop("can_edit", UNSET)

        yaml_parsing_error = d.pop("yaml_parsing_error", UNSET)

        check_definition_model = cls(
            check_name=check_name,
            sensor_name=sensor_name,
            rule_name=rule_name,
            help_text=help_text,
            friendly_name=friendly_name,
            standard=standard,
            custom=custom,
            built_in=built_in,
            can_edit=can_edit,
            yaml_parsing_error=yaml_parsing_error,
        )

        check_definition_model.additional_properties = d
        return check_definition_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/rest/models/metadata/CheckDefinitionModel.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */
package com.dqops.rest.models.metadata;

import com.dqops.checks.DefaultRuleSeverityLevel;
import com.dqops.metadata.definitions.checks.CheckDefinitionSpec;
import com.dqops.metadata.definitions.checks.CheckDefinitionWrapper;
import com.dqops.rules.TargetRuleSeverityLevel;
import com.dqops.utils.docs.generators.SampleStringsRegistry;
import com.dqops.utils.docs.generators.SampleValueFactory;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

import java.util.Objects;

/**
 * Check model that is returned by the REST API. Describes a single unique data quality check.
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "CheckDefinitionModel", description = "Data quality check definition model")
public class CheckDefinitionModel {
    /**
     * Data quality check name.
     */
    @JsonPropertyDescription("Check name")
    private String checkName;

    /**
     * Full sensor name.
     */
    @JsonPropertyDescription("Sensor name")
    private String sensorName;

    /**
     * Full rule name.
     */
    @JsonPropertyDescription("Rule name")
    private String ruleName;

    /**
     * Help text for the check editor.
     */
    @JsonPropertyDescription("Help text that is shown in the check editor that describes the purpose and usage of the check")
    private String helpText;

    /**
     * An alternative check's name that is shown on the check editor.
     */
    @JsonPropertyDescription("An alternative check's name that is shown on the check editor.")
    private String friendlyName;

    /**
     * This is a standard data quality check that is always shown on the data quality checks editor screen.
     * Non-standard data quality checks (when the value is false) are advanced checks that are shown when the user decides to expand the list of checks.
     */
    @JsonPropertyDescription("This is a standard data quality check that is always shown on the data quality checks editor screen. Non-standard data quality checks (when the value is false) are advanced checks that are shown when the user decides to expand the list of checks.")
    private boolean standard;

    /**
     * The severity level (warning, error, fatal) for the default rule that is activated in the data quality check editor when the check is enabled.
     */
    @JsonPropertyDescription("The severity level (warning, error, fatal) for the default rule that is activated in the data quality check editor when the check is enabled.")
    private DefaultRuleSeverityLevel defaultSeverity = DefaultRuleSeverityLevel.error;

    /**
     * True when the check is a custom check or is customized by the user.
     */
    @JsonPropertyDescription("This check has is a custom check or was customized by the user.")
    private boolean custom;

    /**
     * True when this check is provided with DQOps as a built-in check.
     */
    @JsonPropertyDescription("This check is provided with DQOps as a built-in check.")
    private boolean builtIn;

    /**
     * Boolean flag that decides if the current user can update or delete this object.
     */
    @JsonPropertyDescription("Boolean flag that decides if the current user can update or delete this object.")
    private boolean canEdit;

    /**
     * Optional parsing error that was captured when parsing the YAML file.
     * This field is null when the YAML file is valid. If an error was captured, this field returns the file parsing error message and the file location.
     */
    @JsonPropertyDescription("Optional parsing error that was captured when parsing the YAML file. " +
            "This field is null when the YAML file is valid. If an error was captured, this field returns the file parsing error message and the file location.")
    private String yamlParsingError;

    /**
     * Default constructor for CheckModel.
     */
    public CheckDefinitionModel() {
    }

    /**
     * Constructor for a rule model with a given {@link CheckDefinitionWrapper}, custom flag, and built-in flag.
     * Sets the properties of the check based on the given {@link CheckDefinitionWrapper}.
     * @param checkDefinitionWrapper The check definition wrapper to use.
     * @param custom Whether the check is a custom check.
     * @param builtIn Whether the check is a built-in check.
     * @param canEdit The current user can edit the check definition.
     */
    public CheckDefinitionModel(CheckDefinitionWrapper checkDefinitionWrapper, boolean custom, boolean builtIn, boolean canEdit) {
        this.checkName = checkDefinitionWrapper.getCheckName();
        CheckDefinitionSpec checkDefinitionSpec = checkDefinitionWrapper.getSpec();
        this.sensorName = checkDefinitionSpec.getSensorName();
        this.ruleName = checkDefinitionSpec.getRuleName();
        this.helpText = checkDefinitionSpec.getHelpText();
        this.friendlyName = checkDefinitionSpec.getFriendlyName();
        this.yamlParsingError = checkDefinitionSpec.getYamlParsingError();
        this.standard = checkDefinitionSpec.isStandard();
        this.defaultSeverity = checkDefinitionSpec.getDefaultSeverity();
        this.custom = custom;
        this.builtIn = builtIn;
        this.canEdit = canEdit;
    }

    /**
     * Checks whether the CheckModel is equal to a given {@link CheckDefinitionWrapper}.
     * @param checkDefinitionWrapper The CheckDefinitionWrapper to compare.
     * @return true if the rule model is equal to the CheckDefinitionWrapper, false otherwise.
     */
    public boolean equalsBuiltInCheck(CheckDefinitionWrapper checkDefinitionWrapper) {
        if (checkDefinitionWrapper == null) {
            return false;
        }

        if (!Objects.equals(checkDefinitionWrapper.getSpec().getSensorName(), this.sensorName)) {
            return false;
        }

        if (!Objects.equals(checkDefinitionWrapper.getSpec().getRuleName(), this.ruleName)) {
            return false;
        }

        if (!Objects.equals(checkDefinitionWrapper.getSpec().getHelpText(), this.helpText)) {
            return false;
        }

        if (!Objects.equals(checkDefinitionWrapper.getSpec().getFriendlyName(), this.friendlyName)) {
            return false;
        }

        if (!Objects.equals(checkDefinitionWrapper.getSpec().getDefaultSeverity(), this.defaultSeverity)) {
            return false;
        }

        if (checkDefinitionWrapper.getSpec().isStandard() != this.standard) {
            return false;
        }

        return true;
    }

    /**
     * Returns the check definition spec object.
     * @return the check definition spec object.
     */
    public CheckDefinitionSpec toCheckDefinitionSpec() {
        CheckDefinitionSpec checkDefinitionSpec = new CheckDefinitionSpec();
        checkDefinitionSpec.setSensorName(this.sensorName);
        checkDefinitionSpec.setRuleName(this.ruleName);
        checkDefinitionSpec.setHelpText(this.helpText);
        checkDefinitionSpec.setFriendlyName(this.friendlyName);
        checkDefinitionSpec.setStandard(this.standard);
        checkDefinitionSpec.setDefaultSeverity(this.defaultSeverity);

        return checkDefinitionSpec;
    }

    public static class CheckDefinitionModelSampleFactory implements SampleValueFactory<CheckDefinitionModel> {
        @Override
        public CheckDefinitionModel createSample() {
            return new CheckDefinitionModel() {{
                setCheckName(SampleStringsRegistry.getCheckName());
                setSensorName(SampleStringsRegistry.getFullSensorName());
                setRuleName(SampleStringsRegistry.getFullRuleName());
                setHelpText(SampleStringsRegistry.getHelpText());
                setDefaultSeverity(DefaultRuleSeverityLevel.error);
                setStandard(false);
                setCustom(true);
                setBuiltIn(false);
                setCanEdit(true);
            }};
        }
    }
}
